python/mlc_llm/quantization/fp8_quantization.py

```diff
@@ -503,3 +426,165 @@ def forward(self, x: nn.Tensor, indptr: nn.Tensor) -> nn.Tensor:  # pylint: disa
+            )
+            # TODO(csullivan) lm head is not sharded. calibration logic needs to change i think
+            if "max" in quantized_linear.runtime:
+                apply_sharding(
+                    tp.ShardScalar(name=shard.name),
+                    f"{shard.name}_q_calibration_scale",
+                    quantized_linear.q_calibration_scale,
+                )
+
+        return quantized_linear
+
+    def forward(self, x: nn.Tensor) -> nn.Tensor:  # pylint: disable=invalid-name
+        if self.runtime == "max-calibration":
+            x, local_scale = nn.op.quantize(
+                x,
+                quantize_dtype=self.activation_dtype,
+                kind="fp8-max",
+                max_int_value=self.max_int_value["x"],
+            )
+
+            local_scale = nn.op.maximum_inplace(local_scale, self.q_calibration_scale)
+            # Calibration done in fp16 mma so convert x and w back to fp16
+            x = nn.op.astype(x, dtype="float16")
+
+            if DataType(self.weight_dtype).type_code in [
+                DataTypeCode.E4M3Float,
+                DataTypeCode.E5M2Float,
+            ]:
+                dequant_func = self.config._dequantize_float8
+                # The below computes w = cast(reinterpret_cast(self.q_weight, float8_e4m3), float16) * self.q_scale
+                w = nn.op.tensor_expr_op(  # pylint: disable=invalid-name
+                    lambda weight, scale: dequant_func(  # pylint: disable=protected-access
+                        weight,
+                        scale,
+                        out_shape=(self.out_features, self.in_features),
+                    ),
+                    name_hint="dequantize_weight",
+                    args=[self.q_weight, self.q_scale],
+                )
+            elif self.weight_dtype == "float16":
+                w = self.q_weight
+        elif self.runtime == "max":
+            local_scale = self.q_calibration_scale
+            x = x / local_scale
+            x = nn.op.astype(x, dtype=self.activation_dtype)
+            w = self.q_weight
+        elif self.runtime == "cast":
+            x = nn.op.astype(x, dtype=self.activation_dtype)
+            w = self.q_weight
+        else:
+            raise NotImplementedError(
+                f"Only max and cast runtimes are supported for FP8 activations, {self.runtime} is unsupported."
+            )
+        w = nn.op.permute_dims(w)
+        x = nn.op.matmul(x, w, out_dtype="float32")
+
+        if self.runtime == "cast":
+            total_scale = 1.0
+        else:
+            if self.runtime == "max" and self.weight_dtype == "e4m3_float8":
+                total_scale = local_scale * self.q_scale
+            else:
+                # for calibration, q_scale is already used to dequantize the weights
+                total_scale = local_scale
+            total_scale = nn.op.astype(total_scale, dtype="float32")
+
+        x *= total_scale
```

Comment:
why don't we use fp16 out_dtype and multiply it with fp16 total_scale? is it because of accuracy concern?

Reply:
Well observed and yes, exactly because of accuracy/numerical issues. 

I tried this first and debugged for a day and found that the FP8 scaling causes a small dynamic range of activations (e.g. [-0.5, 0.5], fp16) to be scaled up to a much larger range ([-448, 448], fp8). As a result, the MMA produces values larger than FP16's max value. So we either need to have both the cast to fp16 AND scale factors fused into the cublas call, or neither. This PR takes the latter approach since the lowering to fuse the scale isn't ready yet.

If we set the matmul output dtype to float16 and don't fuse the scale, some of the outputs of the cublas call are `inf`, after which scaling has no affect. Unfortunately fp16 does not have a satfinite mode.